airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueIdentityMapper.kt

```diff
@@ -8,58 +8,100 @@ import io.airbyte.cdk.load.message.DestinationRecord
+}
+
+class AirbyteValueNoopMapper(
+    override val meta: DestinationRecord.Meta,
+) : AirbyteValueMapper {
+    override fun map(
+        value: AirbyteValue,
+        schema: AirbyteType,
+        path: List<String>,
+        nullable: Boolean
+    ): AirbyteValue = value
+}
+
 open class AirbyteValueIdentityMapper(
-    val meta: DestinationRecord.Meta,
-) {
+    override val meta: DestinationRecord.Meta,
+) : AirbyteValueMapper {
     private fun collectFailure(
         path: List<String>,
         reason: Reason = Reason.DESTINATION_SERIALIZATION_ERROR
     ) {
-        meta.changes.add(DestinationRecord.Change(path.joinToString("."), Change.NULLED, reason))
+        val joined = path.joinToString(".")
+        if (meta.changes.none { it.field == joined }) {
+            meta.changes.add(
+                DestinationRecord.Change(path.joinToString("."), Change.NULLED, reason)
+            )
+        }
     }
 
-    fun map(
+    override fun map(
         value: AirbyteValue,
         schema: AirbyteType,
-        path: List<String> = emptyList()
+        path: List<String>,
+        nullable: Boolean,
     ): AirbyteValue =
         try {
-            when (schema) {
-                is ObjectType -> mapObject(value as ObjectValue, schema, path)
-                is ObjectTypeWithoutSchema ->
-                    mapObjectWithoutSchema(value as ObjectValue, schema, path)
-                is ObjectTypeWithEmptySchema ->
-                    mapObjectWithEmptySchema(value as ObjectValue, schema, path)
-                is ArrayType -> mapArray(value as ArrayValue, schema, path)
-                is ArrayTypeWithoutSchema ->
-                    mapArrayWithoutSchema(value as ArrayValue, schema, path)
-                is UnionType -> mapUnion(value, schema, path)
-                is BooleanType -> mapBoolean(value as BooleanValue, path)
-                is NumberType -> mapNumber(value as NumberValue, path)
-                is StringType -> mapString(value as StringValue, path)
-                is IntegerType -> mapInteger(value as IntegerValue, path)
-                is DateType -> mapDate(value as DateValue, path)
-                is TimeTypeWithTimezone -> mapTimeWithTimezone(value as TimeValue, path)
-                is TimeTypeWithoutTimezone -> mapTimeWithoutTimezone(value as TimeValue, path)
-                is TimestampTypeWithTimezone ->
-                    mapTimestampWithTimezone(value as TimestampValue, path)
-                is TimestampTypeWithoutTimezone ->
-                    mapTimestampWithoutTimezone(value as TimestampValue, path)
-                is NullType -> mapNull(path)
-                is UnknownType -> {
-                    collectFailure(path)
-                    mapNull(path)
+            if (nullable && value is NullValue) {
+                mapNull(path)
+            } else {
+                when (schema) {
+                    is ObjectType -> mapObject(value as ObjectValue, schema, path)
```

Comment:
This block is unchanged, it just went down a level to accommodate the null check

This was necessary because null values were failing the cast